design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+
+## Glossary
+
+- SRE - Site Reliability Engineering
+- SLI - Service Layer Indicator - basically a metric
+- SLO - An assumption about SLI. The SLI should be between the lower bound and the upper bound. [Here’s a good example](https://landing.google.com/sre/workbook/chapters/slo-document/).
+- SLA - Service Level Agreement - what are the SLOs written into an agreement between services (or Red Hat and end users for that matter).
+- Alert - A warning triggered when metrics went beyond a level defined in SLO.
+[Here’s the OpenShift Dedicated SLA (go to page 9)](https://www.redhat.com/licenses/Appendix_4_Online_Global_20200406.pdf).
+- Prometheus - A time-series database. We store metrics there.
+- Grafana - A service that presents diagrams based on metrics from Prometheus.
+- EFK Stack - Elsticsearch, Fluentd and Kibana. A typical log-collecting stack.
+- CLO - Cluster Logging Operator
+- CR - Custom Resource. An API provided by Keycloak Operator for the context of this document.
+
+## Motivation
+
+Managing Keycloak cluster requires involving a dedicated SRE (Site Reliability Engineering) Team. Such a team needs to
+be able to:
+
+* Diagnose if Keycloak cluster is up and running
+* Have sufficient information to diagnose issues
+* Have enough data and metrics to foresee problems that will happen in the near future
+
+Typically, an SRE Team uses dashboards and alerts. Both of them are created based on metrics exported by a Service
+(in this case, the Service is both Keycloak and Keycloak Operator).
+
+Based on metrics, dashboards and alerts, it is possible to define SLA - an agreement between a Service Provider and a
+Service Consumer.
+
+## Scope
+
+What's in the scope of this design:
+- Designing a new Healcheck API. This has been covered in a separate design document. See [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+- Designing and implementing a dashboard for the SRE Team.
+- Defining SLIs and SLOs as well as proposing template SLAs for a managed Keycloak cluster.
+- Defining metrics required by the proposed dashboard as well as designing a mechanism for exporting them to prometheus.
+- Design logging improvements for better error tracking.
+
+What's out of the scope:
+- Traceability - tracking a user across the whole authorization flow. At the moment we don't have any requirements
+in this direction. If we get such a requirement, we will track this progress in a separate epic and design.
+
+## SRE Dashboard
+
+![SRE Dashboard](img/observability-dashboard.png)
+
+NOTE: The diagram implementation and specific charts might change during the implementation phase. The above proposal
+shows only the general idea.
+
+The proposal consists of 3 sections:
+- System health - Enables the SRE Team to quickly tell if Keycloak cluster is up and running. This section consists of:
+  - Readiness and Liveness Probes chart
+  - HTTP 5xx error rate for all major Keycloak Services, including: Admin Console, Account Console, OIDC, SAML
```

Comment:
We probably want a separate metrics for authentication, where "authentication" is the "optional" bit in the middle of an OIDC or SAML authentication request.

There's other services like authz services, token exchange that we probably want to cover as well.

Should probably just be admin or account, not admin console and account console, as the "console" part is not all that interesting since it's just static files.

Reply:
Ok, so let's try to identify the course-grained services we want to expose:
- Admin
- Account
- OIDC
- Token Exchange
- SAML

Token Exchange definitely doesn't belong to "OIDC" as it's part of the oAuth spec. So it needs to be separate.

SAML and authentication has a lot in common conceptually but they use different endpoints. So I think they can be separate. 

Does this split works for you?

Reply:
@slaskawi should we mention User Federation? I'm thinking about how people could identify an issue when users struggle to login with their LDAP credentials for example.

Reply:
@abstractj Yes, that works for me.